core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java

```diff
@@ -0,0 +1,128 @@
+ * stores all pages in memory.
+ */
+@NotThreadSafe
+public class MemoryPageStore implements PageStore {
+  private static final Logger LOG = LoggerFactory.getLogger(MemoryPageStore.class);
+  private final long mPageSize;
+  private final long mCapacity;
+
+  private ConcurrentHashMap<Long, byte[]> mPageStoreMap = null;
+
+  /**
+   * Creates a new instance of {@link MemoryPageStore}.
+   *
+   * @param options options for the buffer store
+   */
+  public MemoryPageStore(MemoryPageStoreOptions options) {
+    mPageSize = options.getPageSize();
+    mCapacity = (long) (options.getCacheSize() / (1 + options.getOverheadRatio()));
+    mPageStoreMap = new ConcurrentHashMap<>();
+  }
+
+  @Override
+  public void put(PageId pageId, byte[] page) throws ResourceExhaustedException, IOException {
+    long pageKey = getKeyFromPageId(pageId);
+    try {
+      byte[] mPage = new byte[page.length];
+      System.arraycopy(page, 0, mPage, 0, page.length);
+      mPageStoreMap.put(pageKey, mPage);
+    } catch (Exception e) {
+      throw new IOException("Failed to put cached data in memory for page " + pageId);
+    }
+  }
+
+  @Override
+  public int get(PageId pageId, int pageOffset, int bytesToRead, byte[] buffer, int bufferOffset)
+      throws IOException, PageNotFoundException {
+    Preconditions.checkArgument(buffer != null, "buffer is null");
+    Preconditions.checkArgument(pageOffset >= 0, "page offset should be non-negative");
+    Preconditions.checkArgument(buffer.length >= bufferOffset,
+        "page offset %s should be " + "less or equal than buffer length %s", bufferOffset,
+        buffer.length);
+    long pageKey = getKeyFromPageId(pageId);
+    if (!mPageStoreMap.containsKey(pageKey)) {
+      throw new PageNotFoundException(pageId.getFileId() + "_" + pageId.getPageIndex());
+    }
+    byte[] mPage = mPageStoreMap.get(pageKey);
+    Preconditions.checkArgument(pageOffset <= mPage.length, "page offset %s exceeded page size %s",
+        pageOffset, mPage.length);
+    int bytesLeft = (int) Math.min(mPage.length - pageOffset, buffer.length - bufferOffset);
+    bytesLeft = Math.min(bytesLeft, bytesToRead);
+    System.arraycopy(mPage, pageOffset, buffer, bufferOffset, bytesLeft);
+    return bytesLeft;
+  }
+
+  @Override
+  public void delete(PageId pageId) throws IOException, PageNotFoundException {
+    long pageKey = getKeyFromPageId(pageId);
+    if (!mPageStoreMap.containsKey(pageKey)) {
+      throw new PageNotFoundException(pageId.getFileId() + "_" + pageId.getPageIndex());
+    }
+    mPageStoreMap.remove(pageKey);
+    LOG.info("Remove cached page, size: {}", mPageStoreMap.size());
+  }
+
+  /**
+   * @param pageId page Id
+   * @return the key to this page
+   */
+  @VisibleForTesting
+  public long getKeyFromPageId(PageId pageId) {
+    // TODO(feng): encode fileId with URLEncoder to escape invalid characters for file name
+    // Key is : MD5(file_id(str)_page_idx(ulong))
```

Comment:
You're able to use the pageId direclty as the key, the extra md5 hashing is unnecessary which might cause conflict in some extreme cases.

See the hashcode and equals method in the PageId class

` @Override
  public int hashCode() {
    return Objects.hashCode(mFileId, mPageIndex);
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) {
      return true;
    }
    if (!(obj instanceof PageId)) {
      return false;
    }
    PageId that = (PageId) obj;
    return mFileId.equals(that.mFileId) && mPageIndex == that.mPageIndex;
  }`

Reply:
Got it. I misunderstood your means of comment before.
Done.